sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -434,7 +434,7 @@ case class DataSource(
           hs.partitionSchema,
           "in the partition schema",
           equality)
-        DataSourceUtils.verifySchema(hs.fileFormat, hs.dataSchema)
+        DataSourceUtils.checkFieldType(hs.fileFormat, hs.dataSchema)
```

Comment:
Doesn't this affect other sources too? not only parquet? After this PR, it won't validate the field names against https://github.com/apache/spark/blob/62cf4d41b12a3a6d94d011a5b76e66ccaa3fed2a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala#L171 anymore. It change the semantic of an API

Reply:
Please update the docs at `FileFormat.supportFieldName`, and PR description with PR title if that's intended.

Reply:
> Please update the docs at `FileFormat.supportFieldName`, and PR description with PR title if that's intended.

updated

Reply:
ah, +1 to narrow it to parquet source only

Reply:
I think for an existing file, it doesn't make sense for spark to check field names, as the file is already there and Spark should try its best to read it.

Reply:
There is actually a case when this vilification is useful. When a user specifies a schema, we can fail fast without reading footers from files. Note that the user specified schema can be different from the actual source schema, e.g. some sources append a null column when there's no matching field like Parquet.

Reply:
if the user-specified schema matches the actual file schema, I think we should still accept it even if it has special chars?

Reply:
We can but I am not sure if these field names would work properly. Especially it would be more problematic when schema evolution is included here as we allow to read special characters but won't write out. Another concern is that PARQUET-1809 mentions something like dots are being used to nested column access, are we sure these file will be read corretly in Parquet without correctness issues? Furthermore, we will be potentially exposed to all kind of these problems from Avro or ORC + unknown sources that implement `FileFormat`.

Reply:
As the UT shows, all case we can use back quote to wrap the column name and query it. Then user can rename it.
So I don't think it's a problem to read it? For dot, parquet also can read it.

Reply:
For Parquet, maybe yes for now. This is actually a bug in Parquet, right? Is it safe to remove this check for all other sources? If we want to fix this, I would treat this specifically for Parquet only for now. I don't think it's safe to assume that column names have different restrictions on both reading and writing. Doing the check in read side is good. It fails fast.

Reply:
I've read https://issues.apache.org/jira/browse/PARQUET-1809 , I don't think it's related to field name restriction. It improves the filter API to not simply split the string by dot to get the column path. Instead, the API allows the caller side to pass `String[]` directly as the column path. It also implies that Parquet has no limitation on the field name.

Most databases can use any string as column name by quoting them, so it's really a weird design choice for a file format to put a limitation on the field name. Anyway, this is out of Spark's control, but I think Spark should not "guess" the name limitation of a file format and enforce it at Spark's side. "Fail fast" is not beneficial enough here, comparing to the risk of not being able to read valid files and stopping users to use Spark.

Reply:
It's not a guess. For ORC case, it leverages ORC schema parsing library. This schema validation logic is from Parquet logic too.

>  It also implies that Parquet has no limitation on the field name.

If that's the case, why not just removing the restriction?

> comparing to the risk of not being able to read valid files and stopping users to use Spark.

I see the risk of allowing this is worse with potentially some corrupt data. 

Can we narrow this fix to Parquet only for now?

Reply:
> For ORC case, it leverages ORC schema parsing library. This schema validation logic is from Parquet logic too.

It seems ORC is the only case that we leverage the library to validate names and can do 100% safe "fail fast". Parquet and Avro just write the checking code manually, w.r.t. some unknown references.

> If that's the case, why not just removing the restriction on field name in both read and write?

I do want to propose it...

> I see the risk of allowing this is worse with potentially some corrupt data.

If Spark can read it, then it's not corrupted. There is nothing wrong with Spark following the format spec and using the standard library to read a file if every other tool can read this file as well.

Anyway, there are only 3 sources implementing this name check: parquet, orc and avro. I'm ok to narrow this fix to parquet first, and fix orc and avro later.

Reply:
Updated